### PR TITLE
[refactor] Update internal representation of metadata to always use metadata dict

### PR DESCRIPTION
**Pull request by smackesey**
### Summary & Motivation

Internal companion PR: https://github.com/dagster-io/internal/pull/5257

Update internal representation of `metadata` from `metadata_entries` to the dict form we have been pushing to users for almost a year.

- Remove `metadata_entries` from all non-public constructor args
- Introduce `MetadataSerializer` custom field serializer to serialize metadata dict to list of metadata entries and deserialize list of metadata entries to metadata dict (for backcompat). Apply to all serdes classes holding `metadata_entries`

**Context**

More than a year ago, it was decided that Dagster's preferred representation of metadata was to be the metadata dict, as opposed to the metadata entries list:

```
### metadata dict (yes)
{
    "foo": MetadataValue.text("bar"),
    "baz": MetadataValue.int(1)
}

### metadata entries list (no)
[
    MetadataEntry("foo", value=MetadataValue.text("bar")),
    MetadataEntry("baz", value=MetadataValue.int(1)),
]
```

Reasons for this transition:

- metadata dict is more user-friendly
- we deprecated the `description` field on `MetadataEntry` so `MetadataEntry` no longer had any greater expressive power over a dict

There was a hitch in this plan though-- `metadata_entries: Sequence[MetadataEntry]` was widely used as an internal representation of metadata throughout the codebase, including on many serdes-whitelisted objects. So we couldn't just replace the metadata entry list everywhere. Instead, we decided on a temporary solution of:

- offer `metadata` kwarg on user-facing APIs to allow them to pass a metadata dict
- internally convert to list of metadata entries
- keep using list of metadata entries everywhere for internal repr

This has a few problems:

- we have a lot of constructors that take either `metadata` or `metadata_entries`. Sometimes `metadata_entries` is marked private (`_metadata_entries`) other times not. It's confusing.
- we inconsistently reexpose metadata as a dict-- but if we do this, and we are using a `MetadataEntry` list for the internal representation, then we have to reconvert to a dict

All in all the current situation is inconsistent and introduces unnecessary complexity to the code.

This PR should be viewed as a stepping stone to the complete elimination of `MetadataEntry` and passing metadata as `List[MetadataEntry]`, which is done upstack: https://github.com/dagster-io/dagster/pull/12769

### How I Tested These Changes

Existing test suite.

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #12724** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12724" target="_blank">Graphite</a> 
    * **PR #12769** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12769" target="_blank">Graphite</a> 
    * **PR #12725** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12725" target="_blank">Graphite</a> 
      * **PR #12726** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12726" target="_blank">Graphite</a> 
        * **PR #12523** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12523" target="_blank">Graphite</a> 
          * **PR #12524** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12524" target="_blank">Graphite</a> 
            * **PR #12553** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12553" target="_blank">Graphite</a> 
              * **PR #12714** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12714" target="_blank">Graphite</a> 
                * **PR #12715** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12715" target="_blank">Graphite</a> 
            * **PR #12716** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12716" target="_blank">Graphite</a> 
            * **PR #12713** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12713" target="_blank">Graphite</a> 
        * **PR #12770** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12770" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/12770?utm_source=stack-comment).

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #12724** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12724" target="_blank">Graphite</a> 
    * **PR #12769** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12769" target="_blank">Graphite</a> 
    * **PR #12725** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12725" target="_blank">Graphite</a> 
      * **PR #12726** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12726" target="_blank">Graphite</a> 
        * **PR #12523** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12523" target="_blank">Graphite</a> 
          * **PR #12524** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12524" target="_blank">Graphite</a> 
            * **PR #12553** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12553" target="_blank">Graphite</a> 
              * **PR #12714** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12714" target="_blank">Graphite</a> 
                * **PR #12715** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12715" target="_blank">Graphite</a> 
            * **PR #12716** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12716" target="_blank">Graphite</a> 
            * **PR #12713** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12713" target="_blank">Graphite</a> 
        * **PR #12770** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12770" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/12770?utm_source=stack-comment).

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #12724** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12724" target="_blank">Graphite</a> 
    * **PR #12769** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12769" target="_blank">Graphite</a> 
    * **PR #12725** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12725" target="_blank">Graphite</a> 
      * **PR #12726** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12726" target="_blank">Graphite</a> 
        * **PR #12523** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12523" target="_blank">Graphite</a> 
          * **PR #12524** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12524" target="_blank">Graphite</a> 
            * **PR #12553** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12553" target="_blank">Graphite</a> 
              * **PR #12714** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12714" target="_blank">Graphite</a> 
                * **PR #12715** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12715" target="_blank">Graphite</a> 
            * **PR #12716** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12716" target="_blank">Graphite</a> 
            * **PR #12713** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12713" target="_blank">Graphite</a> 
        * **PR #12770** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12770" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/12770?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/metadata/__init__.py [hunk 1]
```diff
@@ -105,42 +90,38 @@ def normalize_metadata(
...
+    normalized_metadata: Dict[str, MetadataValue] = {}
+    for k, v in metadata.items():
+        try:
+            normalized_value = normalize_metadata_value(v)
+        except DagsterInvalidMetadata as e:
+            if allow_invalid:
                 deprecation_warning(
                     "Support for arbitrary metadata values",
-                    "1.0.0",
+                    "1.4.0",
```
Comment (dpeng817) [inline python_modules/dagster/dagster/_core/definitions/metadata/__init__.py:107]: 1.3.0?

**Review by dpeng817 — CHANGES_REQUESTED** [review]
I feel like there's some actual new content buried within this refactor and I'm finding it all a bit hard to parse - as I understand it there are a few distinct pieces here: 
- Refactoring metadata fields to no longer produce metadata entries
- Adding a field serializer that serializes metadata dictionaries as lists of metadata entries
- Changing out all arguments/properties that use metadata_entries to use metadata

In at least some of these examples, I think these count as breaking changes, specifically in the cases where there is no alternative? Such as `SourceAsset` - so we couldn't merge that for example immediately.

Can we break this PR up into those constituent parts?  Lmk your thoughts

**Comment by smackesey** [comment]
>I think these count as breaking changes

Unless I've made a mistake, I don't believe there are any breaking changes in this PR. I concentrated them all in an upstack PR to be separately evaluated.

re `SourceAsset`, prior to this PR there is no public API for passing in `MetadataEntry`. And though many classes currently store metadata as `Sequence[MetadataEntry]`, it's not publicly exposed.

In places where `MetadataEntry` _can_ currently be passed in publicly, in this PR it is still accepted (as in e.g. `AssetMaterialization`)-- the storage format just changes to dict.

**Review by dpeng817 — COMMENTED** [review]
Yea what you're saying makes sense. I think with the additional context it's more parseable - only thing I wonder is if we should potentially keep MetadataEntry around in some code graveyard and add tests that ensure that the format of `MetadataFieldSerializer` matches with that API? As it is, I guess my only request would be tests for `MetadataFieldSerializer` in general

File: python_modules/libraries/dagster-wandb/dagster_wandb/io_manager.py [hunk 2]
```diff
@@ -180,7 +180,7 @@ def _upload_artifact(self, context: OutputContext, obj):
         with self.wandb_run() as run:
             parameters = context.metadata.get("wandb_artifact_configuration", {})  # type: ignore
 
-            serialization_module = parameters.get("serialization_module", {})  # type: ignore
```
[the remark this answers is not in the document]
Comment (smackesey) [reply python_modules/libraries/dagster-wandb/dagster_wandb/io_manager.py:183]: yes it's no longer needed

**Comment by smackesey** [comment]
@dpeng817 Added a field serializer test in `general_tests/test_backcompat.py`

**Review by dpeng817 — APPROVED** [review]
Nice
